src/libos/src/net/socket/unix/stream/stream.rs

```diff
@@ -273,53 +303,76 @@ impl Info {
     }
 }
 
+/// The listener status of a stream unix socket.
+/// It contains a channel holding incoming connections.
+/// The nonblocking status of the reader end keeps the same with the socket.
+/// The writer end is always non-blocking. The connect function returns
+/// ECONNREFUSED rather than block when the channel is full.
 pub struct Listener {
-    channel: Channel<Endpoint>,
-    nonblocking: AtomicBool,
+    channel: RwLock<Channel<Endpoint>>,
 }
 
 impl Listener {
-    pub fn new(capacity: usize) -> Result<Self> {
+    pub fn new(capacity: usize, nonblocking: bool) -> Result<Self> {
         let channel = Channel::new(capacity)?;
-        // It may incur blocking inside a blocking if the channel is blocking. Set the channel to
-        // nonblocking permanently to avoid the nested blocking. This also results in nonblocking
-        // accept and connect. Future work is needed to resolve this blocking issue to support
-        // blocking accept and connect.
-        channel.set_nonblocking(true);
-        /// The listener is blocking by default
-        let nonblocking = AtomicBool::new(true);
+        channel.set_producer_nonblocking(true);
+        channel.set_consumer_nonblocking(nonblocking);
 
         Ok(Self {
-            channel,
-            nonblocking,
+            channel: RwLock::new(channel),
         })
     }
 
-    pub fn push_incoming(&self, stream_socket: Endpoint) {
-        self.channel.push(stream_socket);
+    pub fn capacity(&self) -> usize {
+        let channel = self.channel.read().unwrap();
+        channel.capacity()
+    }
+
+    pub fn resize(&self, capacity: usize) {
+        if self.capacity() == capacity {
+            return;
+        }
+
```

Comment:
Since `pop_incoming` is a blocking operation, the acquire reader lock may block the resize method for an indefinite time, which means a `listen` method (leading to `resize`) on a listening socket may be blocked by a concurrent `accept` for an indefinite time.

I don't think this issue is much of a big problem since calling `listen` more than once to adjust backlog is very rare (I don't even know about this usage before). So I think can add a TODO here to document the problem and maybe print a warning, but not fix the problem in this commit.

Reply:
done